### ISSUE TITLE
remove storage getters`

### ISSUE THREAD

**Issue by kianenigma** [paritytech/substrate#13259]
... or for now mark them as deprecated, and just remove all usage in substrate.

**Comment by bkchr** [comment]
Which storage getters? You mean the `get(fn something)` attribute?

**Comment by bkchr** [comment]
If yes, what you expect as replacement? Just a normal function?

**Comment by kianenigma** [comment]
Yes, manually implemented if needed. `getters` are one of the last pieces of "magic code" that FRAME macros generate and the less of that we have, the better.

**Comment by kianenigma** [comment]
@muraca will work on this.

**Comment by shawntabrizi** [comment]
I have also found it easier to audit storage access by searching for `StorageName::<T>::`. This getter pattern can obfuscate that.

I think the world would be better without it.

**Comment by ruseinov** [comment]
> I have also found it easier to audit storage access by searching for `StorageName::<T>::`. This getter pattern can obfuscate that.
> 
> I think the world would be better without it.

Arguably it's just a matter of adding `#[pallet::getter(` pattern into the mix, but tbh I totally agree that the micro-convenience of this is probably not worth it. If there's too much bloat - implementing this is quite easy. The cool thing about this macro though is that allows for not supplying the type, because even if we implement this by hand, it'd still need to be something like `get_blah::<T>`.  That said, visibility is more important.

**Comment by muraca** [comment]
Hi folks! I have two questions:
1. What should the deprecation message look like? At the moment, I put "#[pallet::getter] is deprecated and will be unavailable in future versions.".
2. Should I manually implement getters for all the Storage items which have a `pub` identifier? I think they're set as public to be reachable from other pallets. Inside the pallet itself, I'll use the `StorageName::<T>::` syntax anyway.

**Comment by bkchr** [comment]
> I have also found it easier to audit storage access by searching for `StorageName::<T>::`

That will not bring you that much as you can also do `<StorageName<T>>::get()` :P

**Comment by kianenigma** [comment]
> > I have also found it easier to audit storage access by searching for `StorageName::<T>::`
> 
> That will not bring you that much as you can also do `<StorageName<T>>::get()` :P

Needing to keep an eye for 2 patterns is still less than 3 🙈

**Comment by bkchr** [comment]
I don't wanted to argue against this issue :P

**Comment by kianenigma** [comment]
I have been having a lot of second doubt about this given some recent discussions elsewhere.

I can see to arguments against getters: 

1. Code that is generated, which you don't write yourself, going against "explicitness". This also means more things that we have to "teach".
2. Harder audit. I think it is a mental help for reading FRAME code to actually distinguish between a function that reads storage with a slightly different syntax, directly referring the storage `struct` rather than a function.  

The former is not a particularly strong argument, and if you really want to go down that rabbit whole, there is a lot of such code in FRAME and it is futile to pursue explicit-ness in an absolutist manner. 

The latter is a stronger argument, but I am having a hard time convincing myself that is is the right fit for _everyone_. If I/some Parity developers don't like getters, we can simply not use them internally. Reviewing external code would still be a pain for me, but that would be my problem. But, _IFF_ we could gather the data to argue that majority of FRAME developers think similar to me, then we could argue that the readability of code improves across the ecosystem if we agree on one paradigm.

**Comment by gavofyork** [comment]
FWIW I originally introduced this and I no longer use it prefering instead consistency of the `::get()` API. I'm happy to mark as deprecated and remove usages.

**Comment by KiChjang** [comment]
I believe that there is a use case for getters, and that is that of publicity -- you may have a storage item that is private or crate-public, but you also want people to be able to publicly get the value under the storage item, so you expose the getter. In short, the `getter` attribute is useful when you only want to expose the read functionality, but nothing else.

**Comment by muraca** [comment]
@KiChjang for that you can manually implement a getter function.

**Comment by xlc** [comment]
> @KiChjang for that you can manually implement a getter function.

That's not a good argement. Many languages offers syntax sugar to generate getter and you can't just say those features should be removed because it is possible to manually write the equivalent code. I don't want to write those code.

**Comment by KiChjang** [comment]
Indeed, I don't see a good reason to remove it just because there's a verbose syntax to write the exact same functionality. The verbosity here adds no value and instead adds another possibility to make copy and paste errors, because I'm very certain that users would just not manually write a getter function every time they need one -- they'd just copy it from another pallet.

**Comment by gavofyork** [comment]
@xlc Having the feature does not come for free. Compared to using the `Item::get()` API (rather than `Self::get_item()`) It's extra code both in Substrate (to support the macro) and in pallets (the attribute macro).

So the question is not "people are using it; why remove it?" (though admittedly this can reasonably place a time limit on the deprecation timeline). It's instead "what is the benefit of the feature, and what cost does it bring?"; in this case the only real benefit appears to be the ability to have a public getter with all other access of the item (e.g. mutating) being restricted.

The cost is code inconsistency: 
- the same operation can be done in two ways - a pallet might use `Item::<T>::get()`, `Self::get_item()` or a mix of the two;
- you might assume that `Self::get_item()` is exactly equivalent to `Item::<T>::get()` however this is not necessarily the case. This might lead to errors in review or audit.

As well as additional complexity in the pallet and substrate to support the feature.

Personally, I'm unconvinced that the benefit outweighs the cost. Perhaps there are other unspecified benefits (if so it would be good to bring these up). If the feature is determined to be not worth the cost, then the points of deprecation and final removal (as well as a possible compatibility path like perhaps a `frame_compat` crate with a macro to do something similar) should be properly discussed and have a timeline of length corresponding to the community's dependence on the feature.

**Comment by xlc** [comment]
Compatibility path will be very helpful. 

I can live without this feature. It is mostly I prefer to not touch code that aren’t broken. Proper deprecation is a must and more discussion with wider audiences will be great.

**Comment by KiChjang** [comment]
I think the magic here is that only `#[pallet::storage]` has this special attribute where we have this extra syntax and parsing for the `#[pallet::getter]` attribute, which isn't standard Rust syntax _and_ is yet another cognitive burden for memorizing the seemingly magical syntax of FRAME.

I actually believe generating code via an attribute _prevents_ errors, as there's no possibility for the functions `<Item<T>>::get()` and `Self::get_item()` to  diverge if the latter's function body is defined as a call to the former. That said, I also noticed that we're in Rust, and standard Rust projects have been living without special syntax for getters/setters -- instead, what happens is that it's done by convention: getters are named after the field/item that they're obtaining the data from (i.e. `item()`); fallible getters are prefixed with a `get_*` (i.e. `get_item() -> Option<T>`); and setters are prefixed with `set_*` (i.e. `set_item()`).

In general, I am in favor of not diverging too much from the Rust syntax, so I can too live without this feature. The only one thing that I am not too sure about is whether this approach in FRAME syntax is the best, since an argument can be made to improve upon the standard Rust syntax in FRAME.

**Comment by kianenigma** [comment]
> I can live without this feature. It is mostly I prefer to not touch code that aren’t broken. Proper deprecation is a must and more discussion with wider audiences will be great.

We can make this be a pilot for learning a mature deprecation path. I know that this has also been discussed internally recently, @ggwpez might be able to chime in. At the moment, our communication is solely based on PRs and the `note_worthy` label. Usually, we mark a feature as deprecated in substrate and remove usage within our codebase to satisfy the CI. Other teams will see the warning and will have time to remove it. Within a few months, we actually remove the code. At the moment, I can see two issues in this pipeline: 

- https://github.com/paritytech/substrate/issues/13260
- https://github.com/paritytech/substrate/issues/12248

What else we can do? 

One thing that comes to mind is making sure such PRs make it into the Substate Developer Newsletter, maintained by @sacha-l, although I am sure he already takes all the PRs that are `note-worthy` labeled. 

--- 

Moreover, while we have a few experts in this thread, it would be good to get a few more opinions. This could be a good task for @juangirini and @aaronbassett to take over: find the right means to do some kind of soft-voting within FRAME developers in the ecosystem to see what the sentiment is.

**Comment by juangirini** [comment]
> Yea I think we can move the creation of this deprecation-process into its own issue @kianenigma, or?

@ggwpez @kianenigma @aaronbassett I moved this conversation to a separate issue #13976
